src/Utilities/CssDocument.php

```diff
@@ -0,0 +1,283 @@
+    }
+
+    /**
+     * Extracts `@import` and `@charset` rules from the supplied CSS.  These rules must not be preceded by any other
+     * rules, or they will be ignored.  (From the CSS 2.1 specification: "CSS 2.1 user agents must ignore any '@import'
+     * rule that occurs inside a block or after any non-ignored statement other than an @charset or an @import rule."
+     * Note also that `@charset` is case sensitive whereas `@import` is not.)
+     *
+     * @param string $css CSS with comments removed
+     *
+     * @return array{0: string, 1: string}
+     *         The first element is the CSS with the valid `@import` and `@charset` rules removed.  The second element
+     *         contains a concatenation of the valid `@import` rules, each followed by whatever whitespace followed it
+     *         in the original CSS (so that either unminified or minified formatting is preserved); if there were no
+     *         `@import` rules, it will be an empty string.  The (valid) `@charset` rules are discarded.
+     */
+    private function extractImportAndCharsetRules(string $css): array
+    {
+        $possiblyModifiedCss = $css;
+        $importRules = '';
+
+        while (\preg_match(self::AT_CHARSET_OR_IMPORT_RULE_PATTERN, $possiblyModifiedCss, $matches)) {
+            [$fullMatch, $atRuleAndFollowingWhitespace, $atRuleName] = $matches;
+
+            if (\strtolower($atRuleName) === 'import') {
+                $importRules .= $atRuleAndFollowingWhitespace;
+            }
+
+            $possiblyModifiedCss = \substr($possiblyModifiedCss, \strlen($fullMatch));
+        }
+
+        return [$possiblyModifiedCss, $importRules];
+    }
+
+    /**
+     * Extracts at-rules with nested statements (i.e. a block enclosed in curly brackets) from the supplied CSS, with
+     * the exception of conditional group rules.  Currently, `@media` is the only supported conditional group rule;
+     * others will be extracted by this method.
+     *
+     * These rules can be placed anywhere in the CSS and are not case sensitive.
+     *
+     * `@font-face` rules will be checked for validity, though other at-rules will be assumed to be valid.
+     *
+     * @param string $css CSS with comments, `@import` and `@charset` removed
+     *
+     * @return array{0: string, 1: string}
+     *         The first element is the CSS with the at-rules removed.  The second element contains a concatenation of
+     *         the valid at-rules, each followed by whatever whitespace followed it in the original CSS (so that either
+     *         unminified or minified formatting is preserved); if there were no at-rules, it will be an empty string.
+     */
+    private function extractNonConditionalAtRules(string $css): array
+    {
+        $possiblyModifiedCss = $css;
+        $atRules = [];
+
+        while (\preg_match(self::NON_CONDITIONAL_AT_RULE_PATTERN, $possiblyModifiedCss, $matches)) {
+            [$fullMatch, $atRuleName] = $matches;
```

Comment:
Please let's keep the type annotation above this line (copied from `CssInliner`):
```php
** @var array<int, string> $matches */
```

Reply:
I think I had cut cut and pasted the code before this was added in.  I've restored it.  Interesting that Psalm neither complains that it is unnecessary nor if it is absent.